src/main/java/build/buildfarm/instance/memory/MemoryInstance.java

```diff
@@ -77,8 +78,8 @@ public MemoryInstance(String name, MemoryInstanceConfig config) {
     this(
         name,
         config,
-        /*contentAddressableStorage=*/ new HashMap<Digest, ByteString>(),
-        /*actionCache=*/ new HashMap<Digest, ActionResult>(),
+        /*contentAddressableStorage=*/ new ConcurrentHashMap<Digest, ByteString>(),
```

Comment:
Consider declaring the field and param as ConcurrentMap to prevent other callers from passing in a plain map.